### PR TITLE
[oneDAL] Update array spec

### PR DESCRIPTION
**Pull request by rlnx**
- Change wording for some array methods
- Update data ownership requirements
- Add implementation notes 

Link to generated HTML: https://rlnx.github.io/oneapi-spec/bugfix-1.2-array-new/data_management/array.html

### CODE REVIEW

**Comment by rlnx** [comment]
@outoftardis, please have a look

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 1]
```diff
@@ -1,276 +1,331 @@
...
-    /// @param ref   The array which shares ownership structure with created
-    ///              one.
-    /// @param data  Mutable or immutable unmanaged pointer hold by
-    ///              created array.
-    /// @param count The number of elements of type $Data$ in the $data$.
+    /// @param ref   The reference array which shares ownership with the created one.
+    /// @param data  The mutable or immutable unmanaged pointer hold by the created array.
+    /// @param count The number of elements of type ``Data`` in the ``data``.
     ///
-    /// @pre  :expr:`std::is_same_v<ExtData, const Data> || std::is_same_v<ExtData, Data>`
```
Comment (michael-smirnov) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:146]: Why this precondition is removed?

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 2]
```diff
@@ -1,276 +1,331 @@
...
+    /// copies content of the immutable memory block into it. The array manages lifetime of the
+    /// allocated mutable memory block. Returns reference to the same array instance.
     ///
     /// @param queue The SYCL* queue object.
-    /// @param alloc The kind of USM to be allocated
+    /// @param alloc The kind of USM to be allocated.
     ///
-    /// @post :expr:`has_mutable_data() == true`
+    /// @post :literal:`has_mutable_data() == true`
     array& need_mutable_data(sycl::queue& queue,
```
Comment (michael-smirnov) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:227]: I think it's worth marking it somehow that this is deprecated

Comment (michael-smirnov) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:227]: What about `need_mutable_data(void)`?

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 3]
```diff
@@ -1,276 +1,331 @@
...
     /// @param queue The SYCL* queue object.
-    /// @param count The number of elements of type $Data$ to allocate memory for.
-    /// @param alloc The kind of USM to be allocated
+    /// @param count The number of elements of type ``Data`` to allocate memory for.
+    /// @param alloc The kind of USM to be allocated.
+    ///
+    /// @pre :literal:`count > 0`
+    /// @post :literal:`get_count() > count`
+    /// @post :literal:`get_count() > count`
     void reset(const sycl::queue& queue,
```
Comment (michael-smirnov) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:257]: What about deprecation & similar changes like in `need_mutable_data`?

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 4]
```diff
@@ -1,276 +1,331 @@
...
+    /// @param ref   The reference array which shares ownership with the created one.
+    /// @param data  The mutable or immutable unmanaged pointer hold by the created array.
+    /// @param count The number of elements of type ``Data`` in the ``data``.
+    ///
+    /// @pre :literal:`data != nullptr`
+    /// @pre :literal:`count > 0`
+    /// @post :literal:`get_count() == count`
+    /// @post :literal:`get_data() == data`
+    template <typename RefData, typename ExtData>
+    void reset(const array<RefData>& ref, ExtData* data, std::int64_t count);
```
Comment (michael-smirnov) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:328]: What about `get_queue()` method?

**Review by michael-smirnov — COMMENTED** [review]
General: what caused such a massive rephrase of method descriptions?

File: source/elements/oneDAL/include/oneapi/dal/array.hpp (hunk 2 above)
Comment (rlnx) [reply source/elements/oneDAL/include/oneapi/dal/array.hpp:227]: We haven't approved that change

**Comment by rlnx** [comment]
@michael-smirnov, all those methods you mentioned, still not discussed at Paul's meeting

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 5]
```diff
@@ -1,276 +1,331 @@
...
 class array {
 public:
     using data_t = Data;
 
-public:
-    /// Allocates a new memory block for mutable data, does not initialize it,
-    /// creates a new array instance by passing a pointer to the memory block.
-    /// The array shall own the memory block (for details, see :txtref:`data_ownership_requirements`).
+    /// Creates a new array instance by allocating a mutable memory block. The created array manages
+    /// lifetime of the allocated memory block.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:13]:
```suggestion
    /// the lifetime of the allocated memory block.
```

same change in other places

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 6]
```diff
@@ -1,276 +1,331 @@
...
-    /// An aliasing constructor: creates a new array instance that stores $data$ pointer,
-    /// assigns the pointer to the ownership structure of $ref$ to the new instance.
-    /// Array shall return $data$ pointer as its mutable or immutable block depending
-    /// on the $ExtData$ type.
+    /// Creates a new array instance that shares ownership with the reference array while storing
+    /// pointer to another memory block provided by the user. Lifetime of the user-provided memory
+    /// block is not managed by the created array. One of use cases of this constructor is creation
+    /// of an array with offset, for example, :literal:`array{ other, other.get_data() + offset }`.
+    /// The array created this way, shared ownership with the ``other`` array, but points to its
+    /// data with offset.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:172]:
```suggestion
    /// The array created this way shares the ownership with the ``other`` array but points to its
    /// data with an offset.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 7]
```diff
@@ -1,276 +1,331 @@
...
+    /// The pointer to the immutable memory block.
+    ///
+    /// @invariant :literal:`get_data() != nullptr` if :literal:`get_count() > 0`
+    /// @invariant :literal:`get_data() == get_mutable_data()` if :literal:`has_mutable_data() == true`
     const Data* get_data() const noexcept;
 
-    /// Returns whether array contains :expr:`mutable_data` or not
-    ///
-    /// @invariant :expr:`mutable_data != nullptr` if this returns `true` and :expr:`count > 0`
+    /// Returns whether array contains mutable data or not.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:210]:
```suggestion
    /// Returns whether an array contains mutable data or not.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 8]
```diff
@@ -1,276 +1,331 @@
...
+    ///
+    /// @pre :literal:`count > 0`
+    /// @post :literal:`get_count() == count`
+    /// @post :literal:`has_mutable_data() == true`
     void reset();
 
-    /// Allocates a new memory block for mutable data, does not initialize it,
-    /// creates ownership structure for this block, assigns the structure inside the array.
-    /// The array shall own allocated memory block.
+    /// Releases ownership of the managed memory block and replace it by a newly allocated mutable
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:247]:
```suggestion
    /// Releases the ownership of the managed memory block and replace it by a newly allocated mutable
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 9]
```diff
@@ -1,276 +1,331 @@
...
+    /// @post :literal:`get_count() > count`
+    /// @post :literal:`get_count() > count`
     void reset(const sycl::queue& queue,
                std::int64_t count,
                const sycl::usm::alloc& alloc = sycl::usm::alloc::shared);
 
-    /// Creates the ownership structure for memory block of externally-allocated mutable data,
-    /// assigns input $deleter$ object to it,
-    /// sets :expr:`data` and :expr:`mutable_data` pointers to this block.
+    /// Releases ownership of the managed memory block and replace it by a pointer to
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:261]:
```suggestion
    /// Releases the ownership of the managed memory block and replace it by a pointer to
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 10]
```diff
@@ -1,276 +1,331 @@
...
     void reset(Data* data,
                std::int64_t count,
                Deleter&& deleter,
-               const sycl::vector_class<sycl::event>& dependencies = {});
+               const std::vector<sycl::event>& dependencies = {});
 
-    /// Creates the ownership structure for memory block of externally-allocated immutable data,
-    /// assigns input $deleter$ object to it,
-    /// sets :expr:`data` pointer to this block.
+    /// Releases ownership of the managed memory block and replace it by a pointer to
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:287]:
```suggestion
    /// Releases the ownership of the managed memory block and replace it by a pointer to
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 11]
```diff
@@ -1,276 +1,331 @@
...
+    ///                      and not throw any exceptions.
     ///
-    /// @param data          The immutable memory block pointer to be assigned inside the array
-    /// @param count         The number of elements of type $Data$ into the block
-    /// @param deleter       The object used to free $data$.
-    /// @param dependencies  Events indicating availability of the $data$ for reading or writing.
+    /// @param data         The pointer to externally-allocated memory block.
+    /// @param count        The number of elements of type ``Data`` in the ``data``.
+    /// @param deleter      The object used to free ``data``.
+    /// @param dependencies Events indicating availability of the ``data`` for reading or writing.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:299]:
```suggestion
    /// @param dependencies Events indicating the availability of the ``data`` for reading or writing.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 12]
```diff
@@ -1,276 +1,331 @@
...
-               const sycl::vector_class<sycl::event>& dependencies = {});
-
-    /// Initializes :expr:`data` and :expr:`mutable_data` with data pointer,
-    /// :expr:`count` with input $count$ value, initializes
-    /// the pointer to ownership structure with the one from ref. Array shall
-    /// return $data$ pointer as its mutable block.
+               const std::vector<sycl::event>& dependencies = {});
 
-    /// @tparam RefData    The type of elements in the referenced array.
+    /// Releases ownership of the managed memory block and start managing lifetime of the reference
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:312]:
```suggestion
    /// Releases the ownership of the managed memory block and starts managing the lifetime of the reference
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 13]
```diff
@@ -1,276 +1,331 @@
...
+    /// array while storing pointer to another memory block provided by the user. Lifetime of the
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:313]:
```suggestion
    /// array while storing the pointer to another memory block provided by the user. The lifetime of the
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 14]
```diff
@@ -1,276 +1,331 @@
...
-    void reset(const array<RefData>& ref, const Data* data, std::int64_t count);
-
-    /// Provides a read-only access to the elements of array.
-    /// Shall not perform boundary checks.
-    const Data& operator[](std::int64_t index) const noexcept;
+    /// @tparam RefData The type of elements in the reference array.
+    /// @tparam ExtData Either ``Data`` or ``const Data`` type.
+    ///
+    /// @param ref   The reference array which shares ownership with the created one.
+    /// @param data  The mutable or immutable unmanaged pointer hold by the created array.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:320]:
```suggestion
    /// @param data  The mutable or immutable unmanaged pointer held by the created array.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 15]
```diff
@@ -1,276 +1,333 @@
...
+    /// @post :literal:`has_mutable_data() == true`
     static array<Data> empty(const sycl::queue& queue,
                              std::int64_t count,
                              const sycl::usm::alloc& alloc = sycl::usm::alloc::shared);
 
-    /// Allocates a new memory block for mutable data, fills it with a scalar value,
-    /// creates a new array instance by passing a pointer to the memory block.
-    /// The array shall own the memory block (for details, see :txtref:`data_ownership_requirements`).
+    /// Creates a new array instance by allocating a mutable memory block and filling its content
+    /// with a scalar value. The created array manages lifetime of the allocated memory block.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:28]:
```suggestion
    /// with a scalar value. The created array manages the lifetime of the allocated memory block.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 16]
```diff
@@ -1,276 +1,333 @@
...
     static array<Data> full(sycl::queue& queue,
                             std::int64_t count,
                             Element&& element,
                             const sycl::usm::alloc& alloc = sycl::usm::alloc::shared);
 
-    /// Allocates a new memory block on mutable data, fills it with zeros,
-    /// creates a new array instance by passing a pointer to the memory block.
-    /// The array shall own the memory block (for details, see :txtref:`data_ownership_requirements`).
+    /// Creates a new array instance by allocating a mutable memory block and filling its content
+    /// with zeros. The created array manages lifetime of the allocated memory block.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:48]:
```suggestion
    /// with zeros. The created array manages the lifetime of the allocated memory block.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 17]
```diff
@@ -1,276 +1,333 @@
...
-    /// for mutable data. It is the responsibility of the calling application to free the memory block
-    /// as the array shall not free it when the reference count is zero.
-    ///
-    /// @param data         The pointer to externally-allocated memory block.
-    /// @param count        The number of elements of type $Data$ in the memory block.
-    /// @param dependencies Events indicating availability of the $data$ for reading or writing.
-    /// @pre :expr:`data != nullptr`
-    /// @pre :expr:`count > 0`
+    /// Creates a new array instance from a pointer to externally-allocated mutable memory block.
+    /// The created array does not manage lifetime of the user-provided memory block. It is the
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:63]:
```suggestion
    /// The created array does not manage the lifetime of the user-provided memory block. It is the
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 18]
```diff
@@ -1,276 +1,333 @@
...
+    /// responsibility of the calling application to deallocate the memory block.
+    ///
+    /// @param data         The pointer to the mutable externally-allocated memory block.
+    /// @param count        The number of elements of type ``Data`` in the memory block.
+    /// @param dependencies Events indicating availability of the ``data`` for reading or writing.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:68]:
```suggestion
    /// @param dependencies Events indicating the availability of the ``data`` for reading or writing.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 19]
```diff
@@ -1,276 +1,333 @@
...
-    /// for immutable data. It is the responsibility of the calling application to free the memory block
-    /// as the array shall not free it when the reference count is zero.
-    ///
-    /// @param data         The pointer to externally-allocated memory block.
-    /// @param count        The number of elements of type $Data$ in the memory block.
-    /// @param dependencies Events indicating availability of the $data$ for reading or writing.
-    /// @pre :expr:`data != nullptr`
-    /// @pre :expr:`count > 0`
+    /// Creates a new array instance from a pointer to externally-allocated immutable memory block.
+    /// The created array does not manage lifetime of the user-provided memory block. It is the
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:81]:
```suggestion
    /// The created array does not manage the lifetime of the user-provided memory block. It is the
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 20]
```diff
@@ -1,276 +1,333 @@
...
+    /// responsibility of the calling application to deallocate the memory block.
+    ///
+    /// @param data         The pointer to the immutable externally-allocated memory block.
+    /// @param count        The number of elements of type ``Data`` in the memory block.
+    /// @param dependencies Events indicating availability of the ``data`` for reading or writing.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:86]:
```suggestion
    /// @param dependencies Events indicating the availability of the ``data`` for reading or writing.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 21]
```diff
@@ -1,276 +1,333 @@
...
+    /// @post :literal:`other.get_count() == 0`
+    /// @post :literal:`other.get_data() == nullptr`
+    /// @post :literal:`has_mutable_data() == false`
     array(array<Data>&& other);
 
-    /// Creates a new array instance which owns a memory block of externally-allocated mutable data.
-    /// The ownership structure shall be created for a block, the input $deleter$
-    /// shall be assigned to it.
+    /// Creates a new array instance from a pointer to externally-allocated mutable memory block.
+    /// The created array manages lifetime of the user-provided memory block. The memory block is
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:117]:
```suggestion
    /// The created array manages the lifetime of the user-provided memory block. The memory block is
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 22]
```diff
@@ -1,276 +1,333 @@
...
     ///
     /// @param queue        The SYCL* queue object.
-    /// @param data         The pointer to externally-allocated memory block.
-    /// @param count        The number of elements of type $Data$ in the memory block.
-    /// @param deleter      The object used to free $data$.
-    /// @param dependencies Events that indicate when $data$ becomes ready to be read or written
+    /// @param data         The pointer to the mutable externally-allocated mutable data.
+    /// @param count        The number of elements of type ``Data`` in the memory block.
+    /// @param deleter      The object used to deallocate ``data``.
+    /// @param dependencies Events that indicate when ``data`` becomes ready to be read or written.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:128]: above you have a different description for this parameter

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 23]
```diff
@@ -1,276 +1,333 @@
...
                    std::int64_t count,
                    Deleter&& deleter,
-                   const sycl::vector_class<sycl::event>& dependencies = {});
+                   const std::vector<sycl::event>& dependencies = {});
 
-    /// Creates a new array instance which owns a memory block of externally-allocated immutable data.
-    /// The ownership structure shall be created for a block, the input $deleter$
-    /// shall be assigned to it.
+    /// Creates a new array instance from a pointer to externally-allocated immutable memory block.
+    /// The created array manages lifetime of the user-provided memory block. The memory block is
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:144]:
```suggestion
    /// The created array manages the lifetime of the user-provided memory block. The memory block is
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 24]
```diff
@@ -1,276 +1,333 @@
...
     ///
     /// @param queue         The SYCL* queue object.
-    /// @param data          The pointer to externally-allocated memory block.
-    /// @param count         The number of elements of type $Data$ in the $data$.
-    /// @param deleter       The object used to free $data$.
-    /// @param dependencies  Events indicating availability of the $data$ for reading or writing.
+    /// @param data          The pointer to the immutable externally-allocated memory block.
+    /// @param count         The number of elements of type ``Data`` in the ``data``.
+    /// @param deleter       The object used to dellocate ``data``.
+    /// @param dependencies  Events indicating availability of the ``data`` for reading or writing.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:155]:
```suggestion
    /// @param dependencies  Events indicating the availability of the ``data`` for reading or writing.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 25]
```diff
@@ -1,276 +1,333 @@
...
-    /// @tparam ExtData    Either $Data$ or $const Data$ type.
+    /// @tparam RefData The type of elements in the reference array.
+    /// @tparam ExtData Either ``Data`` or ``const Data`` type.
     ///
-    /// @param ref   The array which shares ownership structure with created
-    ///              one.
-    /// @param data  Mutable or immutable unmanaged pointer hold by
-    ///              created array.
-    /// @param count The number of elements of type $Data$ in the $data$.
+    /// @param ref   The reference array which shares ownership with the created one.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:179]:
```suggestion
    /// @param ref   The reference array which shares the ownership with the created one.
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 26]
```diff
@@ -1,276 +1,333 @@
...
+    ///
+    /// @pre :literal:`count > 0`
+    /// @post :literal:`get_count() == count`
+    /// @post :literal:`has_mutable_data() == true`
     void reset();
 
-    /// Allocates a new memory block for mutable data, does not initialize it,
-    /// creates ownership structure for this block, assigns the structure inside the array.
-    /// The array shall own allocated memory block.
+    /// Releases the ownership of the managed memory block and replace it by a newly allocated
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:249]:
```suggestion
    /// Releases the ownership of the managed memory block and replaces it by a newly allocated
```

File: source/elements/oneDAL/include/oneapi/dal/array.hpp [hunk 27]
```diff
@@ -1,276 +1,333 @@
...
+    ///                 and not throw any exceptions.
     ///
-    /// @param data         The mutable memory block pointer to be assigned inside the array
-    /// @param count        The number of elements of type $Data$ into the block
-    /// @param deleter      The object used to free $data$.
-    /// @param dependencies Events indicating availability of the $data$ for reading or writing.
+    /// @param data         The pointer to the mutable externally-allocated memory block.
+    /// @param count        The number of elements of type ``Data`` in the ``data``.
+    /// @param deleter      The object used to free ``data``.
+    /// @param dependencies Events indicating availability of the ``data`` for reading or writing.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/array.hpp:274]:
```suggestion
    /// @param dependencies Events indicating the availability of the ``data`` for reading or writing.
```

File: source/elements/oneDAL/source/data_management/array.rst [hunk 28]
```diff
@@ -99,39 +99,83 @@ usage scenario:
...
+
+4. If the array object changes its state from immutable to mutable, the reference count of the
+   ownership structure is decremented.
+
+   - If that count reaches zero, the ownership structure
+     deallocates the immutable memory block and the array destroys the ownership structure. The new
+     instance of ownership structure owning the mutable data block is created and replaces the old one.
+
+   - If that count greater than zero, the ownership structure is not destroyed. The new
+     instance of ownership structure owning the mutable data block is created and replaces the old
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/array.rst:171]:
```suggestion
     instance of the ownership structure owning the mutable data block is created, and it replaces the old
```

if you repeat the same sentence for both cases, shouldn't it be just a single sentence?

File: source/elements/oneDAL/source/data_management/array.rst [hunk 29]
```diff
@@ -99,39 +99,83 @@ usage scenario:
...
+     one.
+
+.. TODO: Add note regarding thread safety
+
+.. note::
+   Implementer can choose an arbitrary implementation strategy that satisfy an array requirements.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/array.rst:177]:
```suggestion
   You may choose an arbitrary implementation strategy that satisfies array requirements.
```

File: source/elements/oneDAL/source/data_management/array.rst [hunk 30]
```diff
@@ -99,39 +99,83 @@ usage scenario:
...
 
-3. If an array owns immutable data, ``mutable_data`` shall be ``nullptr``.
+2. If an array represents mutable data, both pointers shall point to the mutable data block.
 
-4. An array shall store the number of elements in the block it owns and shall update
-   the ``count`` property when a new memory block is assigned to the array.
+3. If an array represents immutable data, pointer to the mutable data block shall be ``nullptr``.
 
-5. An array shall store a pointer to the **ownership structure** of the data:
+4. An array shall use shared ownership semantics to manage lifetime of the stored data blocks:
```
[the remark this answers is not in the document]
Comment (rlnx) [reply source/elements/oneDAL/source/data_management/array.rst:114]: No, it should be plural

File: source/elements/oneDAL/source/data_management/array.rst [hunk 31]
```diff
@@ -99,39 +99,87 @@ usage scenario:
...
 
-3. If an array owns immutable data, ``mutable_data`` shall be ``nullptr``.
+2. If an array represents mutable data, both pointers shall point to the mutable data block.
 
-4. An array shall store the number of elements in the block it owns and shall update
-   the ``count`` property when a new memory block is assigned to the array.
+3. If an array represents immutable data, pointer to the mutable data block shall be ``nullptr``.
 
-5. An array shall store a pointer to the **ownership structure** of the data:
+4. An array shall use shared ownership semantics to manage lifetime of the stored data block:
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/array.rst:114]:
```suggestion
4. An array shall use shared ownership semantics to manage the lifetime of the stored data block:
```

**Review by andreynv — APPROVED** [review]

File: source/elements/oneDAL/source/data_management/array.rst [hunk 32]
```diff
@@ -158,22 +165,23 @@ A typical array implementation may be organized in the following way:
...
-   - If that count reaches zero, the ownership structure
-     deallocates the immutable memory block and the array destroys the ownership structure. The new
-     instance of the ownership structure owning the mutable data block is created, and it replaces the old one.
+   - If that count is greater than zero, the ownership structure is not destroyed.
 
-   - If that count is greater than zero, the ownership structure is not destroyed. The new
-     instance of the ownership structure owning the mutable data block is created, and it replaces the old
-     one.
+5. If a copy of the array object is created, the reference count of the ownership structure is
+   incremented and a pointer to the same ownership structure assigned to the created copy.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/array.rst:183]:
```suggestion
   incremented and a pointer to the same ownership structure is assigned to the created copy.
```

File: source/elements/oneDAL/source/data_management/array.rst [hunk 33]
```diff
@@ -129,7 +129,14 @@ The array shall satisfy the following requirements on managing the memory blocks
      construction. If no deleter object provided, an array calls the default deallocating
      function that corresponds to the internal memory allocation mechanism.
 
-5. An array object may own no data. An array like this is called **zero-sized**:
+5. If the managed pointer to the data block is replaced by the other pointer via ``reset()``, an
+   array releases the ownership of the managed pointer and start managing lifetime of the data block
+   represented by the other pointer.
```
Comment (outoftardis) [inline source/elements/oneDAL/source/data_management/array.rst:134]:
```suggestion
5. If a managed pointer to the data block is replaced by another pointer via ``reset()``, the
   array that managed the pointer releases the ownership of it and starts managing the lifetime of the data block
   represented by the other pointer.
```

**Review by outoftardis — APPROVED** [review]

**Pull request merged**